components/pager.js

```diff
@@ -0,0 +1,181 @@
+ * range and fewer than two page numbers exist on one side of the cluster. When the current page
+ * approaches the ends of the page range, more than two visible pages appear in the cluster on the
+ * side facing the near end of the page range. This allowance keeps the width of the entire Pager
+ * component consistent regardless of the current page number, so that the previous/next buttons
+ * don't shift around horizontally.
+ */
+const Pager = ({ currentPage, totalPages, onClick, className }) => {
+  // Create the array of 1-based page numbers, with page 0 to represent an ellipsis before the
+  // current page and page -1 to represent an ellipsis after the current page. No real difference
+  // between these two ellipsis values, but Pager uses distinct values so we don't have duplicate
+  // React keys. No point memoizing `pageNumbers` as it needs recalculating when any prop changes.
+  let pageNumbers;
+  if (totalPages <= 9) {
+    // A total page count of nine or fewer has no ellipses -- just a straight array of
+    // sequential numbers.
+    pageNumbers = _.range(1, totalPages + 1); // _.range ends at max - 1
+  } else {
+    // With more than nine pages, build a cluster of pages around the current page, first by
+    // determining the minimum and maximum page numbers for the cluster. Allow for filling in
+    // extra numbers in the cluster for cases where the current page number approaches the ends
+    // of the page range, and cutting off the page numbers when the current page is within two
+    // pages of either end of the page range.
+    const clusterMin = Math.min(Math.max(1, currentPage - 2), totalPages - 6);
+    const clusterMax = Math.max(Math.min(currentPage + 2, totalPages), 7);
+    const clusterPageNumbers = _.range(clusterMin, clusterMax + 1); // _.range ends at max - 1
+
+    // Determine whether we need an ellipsis before the cluster, or continuous page numbers. If
+    // we need an ellipsis before the cluster, then we need an array with page one and a "0" to
+    // indicate the left ellipsis.
+    const prevFiller = clusterMin >= 4 ? [1, 0] : _.range(1, clusterMin);
+
+    // Determine whether we need an ellipsis after the cluster, or continuous page numbers. If
+    // we need an ellipsis after the cluster, then we need an array with a -1 to indicate the
+    // right ellipsis followed by the last page number.
+    const nextFiller =
+      clusterMax <= totalPages - 3
+        ? [-1, totalPages]
+        : _.range(clusterMax + 1, totalPages + 1);
+
+    // Put together the entire sequence of displayed page numbers and ellipses:
+    // prevFiller -- clusterPageNumbers -- nextFiller
+    pageNumbers = prevFiller.concat(clusterPageNumbers, nextFiller);
+  }
+
+  // Calculate the pixel width of every page number and ellipsis based on the maximum number of
+  // digits in a page number.
+  const pageNumberWidth = 10 + totalPages.toString().length * 10;
```

Comment:
curious how this works? Is the x10 just like 10 pixels per digit? with 10 pixels of buffer (10 + ...)?

Reply:
Yes, 10 pixels per digit plus padding. This prevents the pager width from changing as you move from page to page, preventing the buttons from shifting out from under your mouse while you’re trying to click them.

Reply:
Is the 10 pixels dependent on the font size or anything?

Reply:
Sure, though since we control that I don’t see much of an issue. Even if we make the font bigger without touching this, it should only be off by a little unless we went extreme on that.